crates/wit-bindgen/src/lib.rs

```diff
@@ -2832,11 +2843,22 @@ impl<'a> InterfaceGenerator<'a> {
         for (i, _) in func.params.iter().enumerate() {
             uwrite!(self.src, "arg{}, ", i);
         }
-        uwriteln!(self.src, ")){await_}?;");
 
+        let instrument = if is_async && self.gen.opts.tracing {
+            ".instrument(span.clone())"
+        } else {
+            ""
+        };
+        uwriteln!(self.src, ")){instrument}{await_}?;");
+
+        let instrument = if is_async && self.gen.opts.tracing {
+            ".instrument(span)"
+        } else {
+            ""
+        };
         uwriteln!(
             self.src,
-            "callee.post_return{async__}(store.as_context_mut()){await_}?;"
+            "callee.post_return{async__}(store.as_context_mut()){instrument}{await_}?;"
```

Comment:
This all feels a bit verbose, but it seemed slightly easier to read than the version where I inlined the conditional into the args to the format string. Happy to add the inlined version if that's preferred.